Control in loading status forever

Hi, 

I have imported the managed solution and then I have added the control to a single line of text field on the Contact form.
When I open the contact form, the control is showing as loading with the "Carregando..." written below and it stays like that forever. 
I have checked that there is audit history for the Contact. I don't know where else to check to make this work.
Any idea? 

![Image](https://github.com/user-attachments/assets/dae13eaa-b21b-4583-ab0a-de3dcb435157)

Thank you,
Adrian

I did a bit of investigation and it seems it fails with this error: 
Main.tsx:66  Error fetching data TypeError: Cannot read properties of undefined (reading 'auditid')

it seems like the auditDynamics.AuditRecord is undefined

![Image](https://github.com/user-attachments/assets/4fe55408-7f4b-4962-b01a-17be5d50e2ed)

--- a/PCF/AuditMonitor/ControlManifest.Input.xml
+++ b/PCF/AuditMonitor/ControlManifest.Input.xml
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="utf-8" ?>
 <manifest>
-  <control namespace="pcfs.audit" constructor="AuditMonitor" version="1.0.0" display-name-key="AuditMonitor" description-key="AuditMonitor description" control-type="virtual" >
+  <control namespace="pcfs.audit" constructor="AuditMonitor" version="1.0.1" display-name-key="AuditMonitor" description-key="AuditMonitor description" control-type="virtual" >
     <external-service-usage enabled="false">
     </external-service-usage>
     <property name="name" display-name-key="name_Key" description-key="Primary name column" of-type="SingleLine.Text" usage="bound" required="true" />

--- a/PCF/AuditMonitor/Helper/Utils.ts
+++ b/PCF/AuditMonitor/Helper/Utils.ts
@@ -44,27 +44,34 @@ export class HelperXrm implements IHelperXrm {
                 }).catch((e) => reject(e));
         });
     }
-    ConvertAuditDynamicsToAudit(auditDynamics: IAuditDetails, metadataAttributes?: IAttributesMetadata[]): IAudit {
+    ConvertAuditDynamicsToAudit(auditDynamics: IAuditDetails | undefined, metadataAttributes?: IAttributesMetadata[]): IAudit | undefined {
+        if (!auditDynamics?.AuditRecord) {
+            console.warn("Invalid audit detail returned by Dataverse", auditDynamics);
+            return undefined;
+        }
+
         const dataAttributes: IDataAttributes[] = [];
         const propsAuditDynamics = Object.keys(auditDynamics);
-        const propsOld = propsAuditDynamics.includes("OldValue") ? Object.keys(auditDynamics.OldValue!) : [];
-        const propsNew = propsAuditDynamics.includes("NewValue") ? Object.keys(auditDynamics.NewValue!) : [];
+        const oldValue = auditDynamics.OldValue;
+        const newValue = auditDynamics.NewValue;
+        const propsOld = propsAuditDynamics.includes("OldValue") && oldValue ? Object.keys(oldValue) : [];
+        const propsNew = propsAuditDynamics.includes("NewValue") && newValue ? Object.keys(newValue) : [];
 
         metadataAttributes?.forEach((meta) => {
             if (propsNew.includes(meta.logicalName)) {
                 dataAttributes.push({
                     logicalName: meta.logicalName,
                     displayName: meta.displayName,
-                    oldValue: propsOld.includes(meta.logicalName) ? this.getFormatedValue(propsOld, auditDynamics.OldValue, meta.logicalName) : "",
-                    newValue: propsNew.includes(meta.logicalName) ? this.getFormatedValue(propsNew, auditDynamics.NewValue, meta.logicalName)  : ""
+                    oldValue: propsOld.includes(meta.logicalName) ? this.getFormatedValue(propsOld, oldValue, meta.logicalName) : "",
+                    newValue: propsNew.includes(meta.logicalName) ? this.getFormatedValue(propsNew, newValue, meta.logicalName)  : ""
                 });
             }
             else if (propsNew.includes(`_${meta.logicalName}_value`)) {
                 dataAttributes.push({
                     logicalName: meta.logicalName,
                     displayName: meta.displayName,
-                    oldValue: propsOld.includes(`_${meta.logicalName}_value`) ? auditDynamics.OldValue![`_${meta.logicalName}_value@OData.Community.Display.V1.FormattedValue`] : "",
-                    newValue: propsNew.includes(`_${meta.logicalName}_value`) ? auditDynamics.NewValue![`_${meta.logicalName}_value@OData.Community.Display.V1.FormattedValue`] : ""
+                    oldValue: propsOld.includes(`_${meta.logicalName}_value`) ? oldValue?.[`_${meta.logicalName}_value@OData.Community.Display.V1.FormattedValue`] : "",
+                    newValue: propsNew.includes(`_${meta.logicalName}_value`) ? newValue?.[`_${meta.logicalName}_value@OData.Community.Display.V1.FormattedValue`] : ""
                 });
             }
         });

--- a/PCF/AuditMonitor/Main.tsx
+++ b/PCF/AuditMonitor/Main.tsx
@@ -48,53 +48,56 @@ export function Main(props: IMainProps) {
   const helper = new HelperXrm(props.context);
   React.useEffect(() => {
     const getDataOnload = async () => {
-      try {
-        await getData();
-      } catch (error) {
-        console.error("Error fetching data", error);
-      }
+      await getData();
     }
     getDataOnload();
   }, [props.name]);
 
   const getData = async (val?: string, refresh: boolean = false) => {
+    try {
+      if (refresh)
+        setState({ ...state, isLoading: true })
 
-    if (refresh)
-      setState({ ...state, isLoading: true })
+      const audits: IAudit[] = [];
+      const dataRegister = getDataRegister();
+      const [promiseAudit, promiseMetadataAttributes] = await Promise.all([
+        helper.RetrieveRecordChangeHistoryRequest(dataRegister.entityLogicalName, dataRegister.id),
+        helper.getMetadataAttributes(dataRegister.entityLogicalName)
+      ]);
+      promiseAudit.AuditDetailCollection?.AuditDetails?.forEach((auditDynamics: IAuditDetails) => {
+        const audit = helper.ConvertAuditDynamicsToAudit(auditDynamics, promiseMetadataAttributes);
 
-    const audits: IAudit[] = [];
-    const dataRegister = getDataRegister();
-    const [promiseAudit, promiseMetadataAttributes] = await Promise.all([
-      helper.RetrieveRecordChangeHistoryRequest(dataRegister.entityLogicalName, dataRegister.id),
-      helper.getMetadataAttributes(dataRegister.entityLogicalName)
-    ]);
-    promiseAudit.AuditDetailCollection.AuditDetails.forEach((auditDynamics: IAuditDetails) => {
-      audits.push(helper.ConvertAuditDynamicsToAudit(auditDynamics, promiseMetadataAttributes));
-    });
-    let auditFilter: IAudit[] | undefined = audits;
-    const normalizedSearchText = helper.normalizeText(val);
-    if (val) {
-      auditFilter = audits?.filter(
-        x => x.attributes.some(x => helper.normalizeText(x.displayName).includes(normalizedSearchText)));
+        if (audit) {
+          audits.push(audit);
+        }
+      });
+      let auditFilter: IAudit[] | undefined = audits;
+      const normalizedSearchText = helper.normalizeText(val);
+      if (val) {
+        auditFilter = audits?.filter(
+          x => x.attributes.some(x => helper.normalizeText(x.displayName).includes(normalizedSearchText)));
+      }
+
+
+      setState({
+        dataRegister,
+        auditsDynamics: promiseAudit,
+        metadataAttributes: promiseMetadataAttributes as IAttributesMetadata[],
+        audits: auditFilter,
+        auditsGroup: helper.groupAuditsByDate(auditFilter),
+        txtFilter: val,
+        isLoading: false
+      });
+    } catch (error) {
+      console.error("Error fetching audit data", error);
+      setState({
+        ...state,
+        audits: [],
+        auditsGroup: undefined,
+        txtFilter: val,
+        isLoading: false
+      });
     }
-
-
-    setState({
-      dataRegister: {
-        //eslint-disable-next-line
-        //@ts-ignore
-        id: props.context?.mode.contextInfo.entityId,
-        //eslint-disable-next-line
-        //@ts-ignore
-        entityLogicalName: props.context?.mode.contextInfo.entityTypeName
-      } as IDataRegister,
-      auditsDynamics: promiseAudit,
-      metadataAttributes: promiseMetadataAttributes as IAttributesMetadata[],
-      audits: auditFilter,
-      auditsGroup: helper.groupAuditsByDate(auditFilter),
-      txtFilter: val,
-      isLoading: false
-    });
   }
 
   const getDataRegister = (): IDataRegister => {

--- a/PCF/AuditMonitor/declarations.d.ts
+++ b/PCF/AuditMonitor/declarations.d.ts
@@ -1,0 +1,3 @@
+/// <reference types="powerapps-component-framework" />
+
+declare module "*.css";

--- a/PCF/tsconfig.json
+++ b/PCF/tsconfig.json
@@ -2,5 +2,6 @@
     "extends": "./node_modules/pcf-scripts/tsconfig_base.json",
     "compilerOptions": {
         "typeRoots": ["node_modules/@types"],
+        "ignoreDeprecations": "6.0"
     }
 }

